components/forward_post_modal/forward_post_channel_select.tsx

```diff
@@ -0,0 +1,397 @@
+
+    const getDefaultResults = () => {
+        let options: GroupedOption[] = [];
+
+        const handleDefaultResults = (res: ProviderResults) => {
+            options = [
+                {
+                    label: 'Recent',
+                    options: res.items.filter((item) => item.channel.type !== 'threads').map((item) => {
+                        const {channel} = item;
+                        return {
+                            label: channel.display_name || channel.name,
+                            value: channel.id,
+                            details: channel,
+                        };
+                    }),
+                },
+            ];
+        };
+
+        provider.fetchAndFormatRecentlyViewedChannels(handleDefaultResults);
+        return options;
+    };
+
+    const defaultOptions = useRef<GroupedOption[]>(getDefaultResults());
+
+    const handleInputChange = (inputValue: string) => {
+        return new Promise<GroupedOption[]>((resolve) => {
+            let options: GroupedOption[] = [];
+            const handleResults = (res: ProviderResults) => {
+                const newOptions: Record<string, any> = {};
+                res.items.forEach((item) => {
+                    const {channel} = item;
+                    const option: ChannelOption = {
+                        label: channel.display_name || channel.name,
+                        value: channel.id,
+                        details: channel,
+                    };
+                    let dividerLabel = formatMessage({id: 'suggestion.mention.recent.channels', defaultMessage: 'Recent'});
+                    switch (channel.type) {
+                    case Constants.OPEN_CHANNEL:
+                        dividerLabel = formatMessage({id: 'suggestion.search.public', defaultMessage: 'Recent'});
+                        break;
+                    case Constants.PRIVATE_CHANNEL:
+                        dividerLabel = formatMessage({id: 'suggestion.mention.private.channels', defaultMessage: 'Private Channels'});
+                        break;
+                    case Constants.DM_CHANNEL:
+                        dividerLabel = formatMessage({id: 'suggestion.search.direct', defaultMessage: 'Direct Messages'});
+                        break;
+                    case Constants.GM_CHANNEL:
+                        dividerLabel = formatMessage({id: 'suggestion.search.group', defaultMessage: 'Group Messages'});
+                        break;
+                    }
+
+                    if (!newOptions[channel.type]) {
+                        newOptions[channel.type] = {label: dividerLabel, options: []};
+                    }
+                    newOptions[channel.type].options.push(option);
+                });
+
+                options = Object.values(newOptions);
+            };
+
+            provider.handlePretextChanged(inputValue, handleResults);
+            resolve(options);
```

Comment:
It's a bit of a legacy thing since I wrote the original providers before we used Redux, and I also didn't intend for them to mix results from the client and server. That's also why the providers aren't really connected to Redux properly and why they don't store their results in their either

Reply:
just another reason to write my own logic around that... might take a while though, so stay tuned for changes! :D